Exact Cardinality Count extension

GWphua opened the pull request:
### Description

This PR introduces the `druid-exact-count` extension, providing a new aggregation function for computing the exact distinct count of values within a dimension. Unlike approximate estimators like HyperLogLog, this extension guarantees precision, which is crucial for use cases demanding exact figures.

The patch achieves this by leveraging RoaringBitmap, a data structure optimized for storing and manipulating sets of 64-bit integers with good compression and performance. The extension includes the necessary components for integrating this functionality into Druid's query processing engine.

#### Exact Count Aggregation

The core of this PR is the implementation of an exact count aggregator using RoaringBitmap64.

*   **Behavioral aspects**:
    The aggregator is invoked via the `bitmap64ExactCount` type in native queries or a corresponding SQL function. It's designed to ingest values from any dimension as long as they are of type `long`. 
    *   Configuration is minimal.
    *   Empty inputs or all-null inputs correctly result in a count of 0.
    *   String columns are not supported.

#### Integration Tests
- Changed `./it.sh` to also include `extension-contrib` packages when building Druid image.
- Added IT for druid-exact-count.

#### Differences with [Distinct Count Aggregator](https://druid.apache.org/docs/latest/development/extensions-contrib/distinctcount/)

| Exact Count | Distinct Count |
| -- | -- |
| No prerequisites to configuring hash partition, segment granularity | Prerequisites needed to perform aggregation |
| Works on 64-bit number columns only (BIGINT) | Works on dimension columns (Including Strings, Complex Types, etc) | 


#### Release note
Introduced a new extension `druid-exact-count` which provides an aggregator `BITMAP64_EXACT_COUNT(columnName)` for computing exact distinct counts on numerical columns.

<hr>

##### Key changed/added classes in this PR
 * `Bitmap64ExactCountAggregatorFactory`
 * `Bitmap64ExactCountBuildAggregatorFactory`
 * `Bitmap64ExactCountMergeAggregatorFactory`
 * `Bitmap64ExactCountBuildAggregator`
 * `Bitmap64ExactCountMergeAggregator`
 * `Bitmap64ExactCountBuildBufferAggregator`
 * `Bitmap64ExactCountMergeBufferAggregator`
 * `RoaringBitmap64Counter`
 * `Bitmap64` interface, for extensibility when newer/faster bitmap functions are introduced.
 * `Bitmap64ExactCountBuildComplexMetricSerde`
 * `Bitmap64ExactCountMergeComplexMetricSerde`
 * `Bitmap64ExactCountModule`
 * `Bitmap64ExactCountPostAggregator`
 * `Bitmap64ExactCountSqlAggregator`
 * `it.sh`
 * Docs @ `druid-exact-count.md`
 * Integration Test files

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

FrankChen021 commented on extensions-contrib/druid-exact-cardinality/src/main/java/org/apache/druid/query/aggregation/exact/cardinality/bitmap64/Bitmap64Counter.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.nio.ByteBuffer;
+
+public interface Bitmap64Counter
```
as a core interface in this module, it's better to document methods in this interface.

FrankChen021 commented on extensions-contrib/druid-exact-cardinality/src/main/java/org/apache/druid/query/aggregation/exact/cardinality/bitmap64/sql/Bitmap64ExactCardinalitySqlAggregator.java:
```diff
@@ -0,0 +1,163 @@
+import java.util.Collections;
+import java.util.List;
+
+public class Bitmap64ExactCardinalitySqlAggregator implements SqlAggregator
```
Do we have a test case to cover this SQL layer function?

GWphua replied:
Created test cases under Bitmap64ExactCardinalitySqlAggregatorTest 

Docs can be found in the directory: docs/development/extensions-contrib/druid-exact-cardinality.md

FrankChen021 reviewed (commented):
the implementation itself generally looks good to me. I think we need more document and test cases.

## Document
1. a user doc to state how to use this extension, and the build & merge function provided in this extension
2. SQL function explained and some SQL query examples as well as native query examples
3. how can we use this extension during ingestion phase

## Test Case
I see many UT test cases have been added. I think we need some IT test cases to cover the case from ingestion to query(both native queries and SQL queries)

GWphua commented:
@FrankChen021 Thanks for the feedback, I have added Integration tests by allowing CI to build Druid image with all extension-contrib packages.

Added examples for Bitmap64ExactCardinalityBuild + Bitmap64ExactCardinalityMerge from rolled-up Bitmap64 columns. Please take a look again.

GWphua commented on docs/development/extensions-contrib/druid-exact-cardinality.md:
```diff
@@ -0,0 +1,443 @@
+
+### SQL Query
+
+You can use the `BITMAP64_EXACT_CARDINALITY` function in SQL queries:
```
Yes, as long as the column is a Long column, it can be used without needing to pre-aggregate the column.

GWphua commented on extensions-contrib/druid-exact-cardinality/src/main/java/org/apache/druid/query/aggregation/exact/cardinality/bitmap64/Bitmap64ExactCardinalityObjectStrategy.java:
```diff
@@ -0,0 +1,69 @@
+
+    byte[] bytes = new byte[numBytes];
+    readOnlyBuffer.get(bytes, 0, numBytes);
+    return RoaringBitmap64Counter.fromBytes(bytes);
```
This method will also work with Direct ByteBuffers, in which trying to retrieve the array will result in a 'Unsupported exception'.

FrankChen021 replied:
In the `fromBytes` method, it uses the `deserialize` method to turn the byte array into a bitmap object.
```java
      Roaring64NavigableMap bitmap = new Roaring64NavigableMap();
      bitmap.deserialize(in);
```

The `deserialize` accepts `DataInput` as argument. In Netty, there's `io.netty.buffer.ByteBufInputStream` that wraps a ByteBuffer as `DataInput`. I'm wondering if we can use it or copies the idea from it.

Maybe we can do sth like this?
```java
  if (buffer.hasArray()) {
    return new ByteArrayInputStream(
        buffer.array(), 
        buffer.arrayOffset() + buffer.position(), 
        buffer.remaining()
    );
  } else {
    // wrap the ByteBuffer into DataInput if possible
    // or copy the data from ByteBuffer as byte[]
  }
```

GWphua commented on extensions-contrib/druid-exact-cardinality/pom.xml:
```diff
@@ -0,0 +1,164 @@
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
```
Removed since other modules will not be using any test utilities in this extension.

FrankChen021 commented on extensions-contrib/druid-exact-cardinality/src/main/java/org/apache/druid/query/aggregation/exact/cardinality/bitmap64/Bitmap64ExactCardinalityAggregatorFactory.java:
```diff
@@ -0,0 +1,203 @@
+ */
+public abstract class Bitmap64ExactCardinalityAggregatorFactory extends AggregatorFactory
+{
+  static final int MAX_INTERMEDIATE_SIZE = 1024; // 1 KiB
```
let's put these comments as java doc in the code

FrankChen021 reviewed (commented):
the druid-exact-count name is too generic, can we change it to druid-exact-count-bitmap ?

FrankChen021 reviewed (approved):
LGTM
